File: 1019. Next Greater Node In Linked List.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
	public int[] nextLargerNodes(ListNode head) {
		ArrayList<Integer> list = new ArrayList<>();
		if (head == null) return new int[0];
		int idx = 0;
		Stack<int[]> stk = new Stack<int[]>();
		while (head != null) {
			list.add(0);
			
			while (!stk.isEmpty() && stk.peek()[0] < head.val) {
				int[] curr = stk.pop();
				list.set(curr[1], head.val);
			}
			
			stk.push( new int[]{head.val,idx} );
			
			head = head.next;
			idx++;
		}

		int[] res = new int[list.size()];
		for (int i = 0; i < list.size(); i++) {
			res[i] = list.get(i);
		}
		return res;
	}
}

File: 1019. Next Greater Node In Linked List.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
	vector<int> nextLargerNodes(ListNode* head) {
		vector<int> res;
		if (head == nullptr) return res;
		
		int idx(0), count(0);
		stack<pair<int,int>> monoStk;
		while (head) 
		{
			res.emplace_back(0);
			if (monoStk.empty()) 
			{
				monoStk.push( {head->val,idx} );
				// res.emplace_back(0);
			}
			else 
			{
				while (!monoStk.empty() && monoStk.top().first < head->val) 
				{
					auto [val,idx] = monoStk.top();
					res[idx] = head->val;
					monoStk.pop();
				}
			
				monoStk.push( {head->val,idx } );
			}
			
			head = head->next;
			idx++;
		}
		
		return res;
	}
};